### PR TITLE
Assorted fixes to the test infrastructure and to the lifecycle management

### PR DESCRIPTION
**Pull request by gsmet**
* https://hibernate.atlassian.net/browse/OGM-1113
 * https://hibernate.atlassian.net/browse/OGM-1103
 * https://hibernate.atlassian.net/browse/OGM-1114
 

Ok, so this is mostly Gunnar's work on OGM-1103 with some additions from me.

**OGM-1113** 441031e

No real bug here but as Gunnar thought there was one, we obviously needed some clean up, test case and additional comments.

At least for me, coming with new eyeballs, the test infrastructure is far more clear after this commit but let's discuss it.

See the commit comment for more details.

**OGM-1103** f115e1b

This is a followup of https://github.com/hibernate/hibernate-ogm/pull/711 and https://github.com/hibernate/hibernate-ogm/pull/723 , already largely discussed. It's basically Gunnar's proposal with a proper test case and unrelated changes extracted.

@gunnarmorling I took the liberty to amend your commit so that we have a clean one while keeping attribution.

**OGM-1114**

Noted by me when working on OGM-1064 and by Gunnar when working on OGM-1103 so it's time to tackle this one!

6bbbfd3

Make Byteman usable for OGM tests. I didn't find a better way to do it than to enable for every tests. Note that I already needed it to add counters for OGM-1064 so it was already done.

e446492

We need to set up the event context even for a persist event if JPA is not used. Previously, it was only set up if we were in a JPA context.

### CODE REVIEW

**Comment by DavideD** [comment]
I've added the fix for Neo4j remote, it should work now.

**Comment by DavideD** [comment]
Jenkins, retest this please

**Comment by gsmet** [comment]
Jenkins, retest this please

File: core/src/test/java/org/hibernate/ogm/utils/test/SkipByDatastoreProviderSelfJpaTest.java [hunk 1]
```diff
@@ -6,14 +6,14 @@
...
-import static org.hibernate.ogm.datastore.impl.AvailableDatastoreProvider.NEO4J_EMBEDDED;
-import static org.hibernate.ogm.datastore.impl.AvailableDatastoreProvider.MONGODB;
+import static org.hibernate.ogm.datastore.impl.DatastoreProviderType.CASSANDRA_EXPERIMENTAL;
+import static org.hibernate.ogm.datastore.impl.DatastoreProviderType.COUCHDB_EXPERIMENTAL;
+import static org.hibernate.ogm.datastore.impl.DatastoreProviderType.INFINISPAN;
+import static org.hibernate.ogm.datastore.impl.DatastoreProviderType.MAP;
+import static org.hibernate.ogm.datastore.impl.DatastoreProviderType.EHCACHE;
+import static org.hibernate.ogm.datastore.impl.DatastoreProviderType.REDIS_EXPERIMENTAL;
+import static org.hibernate.ogm.datastore.impl.DatastoreProviderType.NEO4J_EMBEDDED;
+import static org.hibernate.ogm.datastore.impl.DatastoreProviderType.MONGODB;
```
Comment (gsmet) [inline core/src/test/java/org/hibernate/ogm/utils/test/SkipByDatastoreProviderSelfJpaTest.java]: I wanted it to be consistent with `GridDialectType` as they are exactly the same thing.

**Comment by DavideD** [comment]
Jenkins, retest this please

**Comment by DavideD** [comment]
Rebased and merged https://github.com/hibernate/hibernate-ogm/pull/727

**Pull request closed without merging**